### ISSUE TITLE
forgeBSgenomeDataPkgFromNCBI() enhancement: Refine handling of 'circ_seqs' argument

### ISSUE THREAD

**Issue by hpages** [Bioconductor/BSgenomeForge#16]
To do once PR #13 is merged and issue #15 is closed. Make sure to resync your fork once PR #15 is merged.

When using `forgeBSgenomeDataPkgFromNCBI()` to forge a package for an assembly that is not registered, the user is asked to specify the circular sequences via the `circ_seqs` argument. The function then calls `getChromInfoFromNCBI()` internally to retrieve the sequence information for the assembly, and to check that the sequence names suplied via `circ_seqs` are _assembled molecules_ (only an assembled molecule can be circular).

However, when the assembly has no assembled molecule (e.g. GCA_009729545.1), the user is still asked to specify the circular sequences, even though we know that there can't be any.

This could be improved by:
1. Recognizing this situation.
2. Infer the vector of circular sequences (`character(0)`).
3. If the user didn't specify the circular sequences (i.e. `circ_seqs` is `NULL`), then everything is fine. If they did specify them, then they should have given an empty character vector. In other words, in this situation `circ_seqs` can only be `NULL` or a character vector of length 0.

**Comment by kakopo** [comment]
Hi Hervé. I've just opened a pull request for this issue. Please have a look and let me know what you think. Thank you.

**Comment by hpages** [comment]
Thanks!

So this works:
```
.check_sequencerole <- function(circ_seqs)
{
    if (is.null(circ_seqs))
        return(character(0))
    if (length(circ_seqs) > 0)
        stop(wmsg("This assembly does not contain assembled-molecules, and so
                  cannot have any other value of 'circ_seqs' specified other
                  than 'character(0)'"))
    return(circ_seqs)
}
```
but the line in `.get_circseqs()` where you call this is a little surprising:
```
        ...
    } else {
        ## NCBI assembly is **not** registered.
        if (! "assembled-molecule" %in% chrominfo[ , "SequenceRole"])
            circ_seqs <- .check_sequencerole(circ_seqs)
        ...
        ... etc...
        ... etc...
        ... (many things happen between the above call to .check_sequencerole()
        ... and the final return() statement below)
        ... etc...
        ...
        return(circ_seqs)
    }
```
The question is: why do you assign the result returned by `.check_sequencerole()` to a variable (`circ_seqs`)? Is there anything else that you need to do with it?

It's important to realize that, because you are _assigning_ and not _returning_, then the flow of execution will continue and go thru all the other things that follow. It turns out that the only possible value that `.check_sequencerole()` can return is `character(0)`, so `circ_seqs` will actually be able to go thru all these things without problem, until execution reaches the final `return(circ_seqs)`. But it that necessary? It seems to me that when `.check_sequencerole()` returned, then the job was done i.e. there was nothing else to do, so there was no need to continue.

Now about the `.check_sequencerole()` helper itself. It's very simple and it's quite easy to understand what it does, but it could be even simpler. For example:
```
.check_sequencerole <- function(circ_seqs)
{
    if (is.null(circ_seqs) || length(circ_seqs) == 0)
        return(character(0))
    stop(wmsg("This assembly does not contain assembled-molecules, and so
              cannot have any other value of 'circ_seqs' specified other
              than 'character(0)'"))
}
```
This does the same thing but the function has only 2 exit points: the `return()` statement and the `stop()` statement. So it's immediately clear that there are only 2 possible outcomes when calling this function: either it returns `character(0)` (first exit point) or it fails (2nd exit point). The original version has 3 exit points: 2 `return()` statements and one `stop()` statement, but it's not immediately obvious that the 2nd `return()` statement actually returns the same thing as the 1st one (one has to do a little mental computation to figure this out).

Or even simpler:
```
.check_sequencerole <- function(circ_seqs)
{
    if (length(circ_seqs) == 0)
        return(character(0))
    stop(wmsg("This assembly does not contain assembled-molecules, and so
              cannot have any other value of 'circ_seqs' specified other
              than 'character(0)'"))
}
```
because `length(NULL)` is zero :wink:

**Comment by kakopo** [comment]
Hi Hervé!
Thank you for the feedback. I have just updated the PR. Let me know if it is okay, thanks.

**Comment by hpages** [comment]
Hmm...
```
        ...
    } else {
        if (! "assembled-molecule" %in% chrominfo[ , "SequenceRole"])
            circ_seqs <- .check_sequencerole(circ_seqs)
        return(circ_seqs)
        ...
        ...
        return(circ_seqs)
    }

```
Is this what you want to do? This probably breaks a few things. Have you tested this?

**Comment by hpages** [comment]
Also note that whatever code is between the 2 `return()` statements now is _dead code_ i.e. code that will never be executed. This is not a good thing.

**Comment by kakopo** [comment]
Yes, I've seen how this approach was not correct :disappointed_relieved:! I've edited the return statement and it works as expected this time.

**Comment by hpages** [comment]
https://github.com/Bioconductor/BSgenomeForge/pull/22/commits/a3c5dfc68949cd668f53bcd2e075c3111dbdd951 Exactly! :smiley: 

I would argue that the name of your helper function (`.check_sequencerole`) doesn't really help. It's actually misleading because the function does not check the sequence role at all. The only thing it does is check that the user supplied no circular sequences (and raise an error if they did). So maybe a name like `.check_no_circ_seqs` or something like that would help? (Maybe still not super clear but at least not misleading.) Thx!

**Comment by kakopo** [comment]
Done! I've updated the function name. The new name certainly fits in the code context :grinning:.

**Comment by hpages** [comment]
Thanks! I've merged PR #22 so we can close this.

Additionally I made some minor corrections and adjustments and bumped the package version. See commit 4c8e726c46745bbe40c5700a527f67cca25c1f51 Don't forget to resync your fork.
